### ISSUE TITLE
Strong parameters don't permit array of arrays (nested arrays)

### ISSUE THREAD

**Issue by leonardoprg** [rails/rails#23640]
Hey guys,

When I try to permit params contains an Array of Array, strong parameters permit is ignoring the second Array and returning a blank Array.

Example:

> Controller

`params.permit(param1: [param2: [[[:attr, :op, :val]]]])`

> Sent Parameters

`"param1": {"param2": [[{"attr": "1", "op": "<", "val": "2"}]]}`

Rails will ignore the nested Arrays and will return only the 
`{param1: [] }`

**Comment by leonardoprg** [comment]
Problem detected on Strong parameters,

PR incoming

**Comment by maclover7** [comment]
Can you create an executable test script using one of the templates [here](https://github.com/rails/rails/tree/master/guides/bug_report_templates), that would demonstrate the problem you are seeing?

**Comment by leonardoprg** [comment]
@maclover7 

Yes, I created a test file, but **attention please** 
When execute this test through minitest it **won't fail**

but, if you go to a normal rails app and try the same test using postman or curl, it will have a difference behavior and will fail.

> link to the test (doesn't fail)
[https://gist.github.com/leonardoprg/ae6b3a48b880d5630c52](https://gist.github.com/leonardoprg/ae6b3a48b880d5630c52)

> Step by Step to get fail

- rails new test
- cd test
- rails g controller test index
- At routes.rb add `post '/', to: 'test#index'`
- At application_controller.rb comment line `protect_from_forgery with: :exception`
- At test_controller.rb method **index** put `render json: params.permit(param1: [param2: [[[:attr, :op, :val]]]])`

So you can test trough terminal with
`curl -X POST --data '{"param1": {"param2": [[{"attr": "1","op": "<", "val": "2" }]] }}' -H "Content-Type: application/json" http://localhost:3000/`

Please note it is the same code/params at test snippet.

This problem happens at 

[https://github.com/rails/rails/blob/4-2-stable/actionpack/lib/action_controller/metal/strong_parameters.rb#L575-L578](https://github.com/rails/rails/blob/4-2-stable/actionpack/lib/action_controller/metal/strong_parameters.rb#L575-L578)

**I already fixed it** and a new PR is coming.

Regards,

**Comment by diego-silva** [comment]
@leonardoprg You can reproduce the save behaviour as the server on tests using something like:

```ruby
def test_parameters_permit
    params = {"param1": {"param2": [[{"attr": "1","op": "<", "val": "2" }]] }}
    ac_params = ActionController::Parameters.new params

    assert_equal params.to_json, ac_params.permit(param1: [param2: [[[:attr, :op, :val]]]]).to_json
  end
```

**Comment by maclover7** [comment]
:+1: Thank you for fixing the behavior -- please open a PR with a fix and a regression test when you can. :)

**Comment by alexanderfrankel** [comment]
@leonardoprg
any updates on the status of this fix? i saw the PR is still open...
this fix is going to be welcomed with open arms, so thanks!

**Comment by leonardoprg** [comment]
@alexanderfrankel 

Yes the PR is still open, is need someone from rails team review it.

I will comment again on PR and try that someone review it.

**Comment by lgrains** [comment]
I ran into this problem today.  Any updates on when this will be approved and merged?

**Comment by waleedasif322** [comment]
Has this been merged? any idea which version this will be released? Thanks

**Comment by arthurnn** [comment]
@waleedasif322 it has not. We are working on it.
If it gets merged to master, it will be released into rails 5.1, but we don't know yet. I will update this issue once the PR is merged.
Thanks

**Comment by MohammadHabbab** [comment]
Any updates about this 😄 ?

**Comment by diofeher** [comment]
Please update that.

**Comment by Amrrbakry** [comment]
Still waiting for an update.

**Comment by leonardoprg** [comment]
hello guys, the PR is already updated with changes requested by rails team,
just waiting review
please follow: https://github.com/rails/rails/pull/23650

Thanks,

**Comment by jpkalbacher** [comment]
Any updates on this?

**Comment by leonardoprg** [comment]
I think the PR is done, just waiting to someone merge it.

**Comment by danhart** [comment]
Guys, seriously, two years to fix an issue of this magnitude? I can't send my quite normally-structure data to a Rails controller? Surely that's high on the priority list for fixing?

**Comment by akhmel** [comment]
Same for me

**Comment by tilo** [comment]
helloooooo... bump!

**Comment by brtsai** [comment]
bump

**Comment by leonardoprg** [comment]
@rafaelfranca same here https://github.com/rails/rails/pull/23650#issuecomment-367178282
It has been 2 years you guys are passing this PR to another one,  so please instead come here and be ironic with the guys, how about have respectful with the work and time spent to get it working and answer the question, or just review the PR?
if you guys really believe in your words about `hey, come here and contribute to rails, we really enjoy you guys work`, 
I sorry to say, you are doing this wrong.

**Comment by rafaelfranca** [comment]
Sorry but bumping to get attention is also disrespectful, a lot. My give a lot of our time for free to people, and we do that because we like, but nobody likes when people is poking you to get something that you are not ready to give.

And just to be clear I was not being ironic. I was being really direct, bumping just annoy and cause people to avoid to review the PR and will not get anything fixed. I was not being disrespectful either, they are, with the maintainers and with the contributor that spent time to fix the problem.

You don't enter a working kitchen and ask the chef to hurry because you are hungry.

I'm sorry that the review took so long, but as you may have noticed Rails is made by volunteers and they have life, in the same way that the contributors. Maintainers don't poke people to update they PRs or issues in respect to their time and life, so we expect the same respect from the contributors.

Focusing in the solution of the problem, ee are still in doubt about the implementation:

> It just fixes a problem with the nested array pattern, like: [[{ a: :b }], [{ b: :c }]] it does not add a new pattern.

Is not an answer for the question made. The question is: do this change now start to permit something that was not permitted before?

We get that it add a new pattern, but does that means that no existing pattern would change its behavior? Or they would?

To give an example, would `params.permit(foo: [])` have its behavior changed?

**Comment by leonardoprg** [comment]
I think the guys are just discrediting that this PR will be reviewed and merge, once took almost 1 year to have a review.
What I expect from a rails team guy, instead to spent time going here to be "direct" with an irrelevant comment ( looking through code and PR ) is to go to PR and comment, `to review it I need that X answer the question` or something like that.
but that's ok, I think the focus should be to get it merged.

Answering the question, I really don't know,
I have not more rails env setuped on my environment, and have no much time to get it done right now.
so, if you want to close this PR, and allow another one to open a new PR with the code, could do without worries

**Comment by st0012** [comment]
I want to propose another syntax:

We currently don’t have any symbol to indicate if we permit an array or an object. `[]` in strong parameter’s semantics only has one meaning at this point: a list of permitted attributes.  It can not be used to decide the permitted params’ structure.

```ruby
param1 = ActionController::Parameters.new(book: { title: "foo"  })
param1.permit(book: [:title]).to_h 
#=> {"book"=>{"title"=>"foo"}}
param2 = ActionController::Parameters.new(book: [{ title: "foo"  }])
param2.permit(book: [:title]).to_h
#=> {"book"=>[{"title"=>"foo"}]}
```

Here’s how we use an `[]` now, if it contains
- symbols: only permit these attributes
- hashes: only permit hash keys attributes and resolve those attributes’ attributes
- nothing: permit anything/any key

As you can see, it requires names (symbol or hash key) to make it work.  This means `[[:title]]` breaks the rule by not providing names.  And since `books: [:title]` means permitting a single object/a collection of objects with title. I say we follow that rule, which is providing a name to declare what we’re doing. This means using something like `books: {[] => [:title]}` to declare that we permit an array of a single object/a collection of objects with title. So it will permit both of

```ruby
{
  "books" => [{ title: "foo" }]
}
```

And 
```ruby
{
  "books" => [
    [{ title: "foo" }]
  ]
}
```


So back to the issue example: `{"param1": {"param2": [[{"attr": "1","op": "<", "val": "2" }]] }}`

I think we should use `permit(param1: [param2: {[] => [:attr, :op, :val]}])`

And it’ll be:

```ruby
params = {
  "param1": {
    "param2": [
      [{"attr": "1","op": "<", "val": "2" }]
    ] 
  }
}

params.permit(param1: [param2: {[] => [:attr, :op, :val]}])
```

**Comment by st0012** [comment]
@leonardoborges Do you still have this issue? Would you mind take a look at my proposal above?

**Comment by mahfuz10** [comment]
when you want to permit `books` **array of hash**:

```
{
  "library": {
    "name" : "my library",
    "books": [{ title: "foo", author: "John" }, { title: "foo", author: "James" }]
  } 
}
```
This code should work:
```
params.require(:library).permit(:name, books: [:title, :author])
```
**Important:** You need to write hash index (`:title, :author`) in `books` array in `permit`.
This way you can permit more nested attributes.

```
params.require(:order).permit(:recipient_name, :recipient_phone,
        order_items_attributes: [:id, :product_name, :_destroy, product_addons: [:name, :option, :price]])
```
Here `product_addons` is an **Array of Hash**. I used this code in a **Rails 5** project for nested form and it works!

Thanks @st0012

**Comment by tke578** [comment]
Hi All,

**Comment by tke578** [comment]
Hi all, I'm curious why I am able to create a nested array but during the rails controller request, the nested array gets flattened? ```<ActionController::Parameters {"components"=>[["abc", "123"], ["def", "456"]]} permitted: true>```

**Comment by kopecmark** [comment]
> Hi all, I'm curious why I am able to create a nested array but during the rails controller request, the nested array gets flattened? `<ActionController::Parameters {"components"=>[["abc", "123"], ["def", "456"]]} permitted: true>`

Running into the same issue with geoJSON schema POST requests.
